selene-lib/src/rules/high_cyclomatic_complexity.rs

```diff
@@ -0,0 +1,389 @@
+                        }
+
+                        ast::Field::NoKey(expression) => {
+                            complexity = count_expression_complexity(expression, complexity);
+                        }
+
+                        _ => {}
+                    }
+                }
+
+                complexity
+            }
+
+            ast::Value::Var(ast::Var::Expression(var_expression)) => {
+                for suffix in var_expression.suffixes() {
+                    #[cfg_attr(
+                        feature = "force_exhaustive_checks",
+                        deny(non_exhaustive_omitted_patterns)
+                    )]
+                    match suffix {
+                        ast::Suffix::Index(ast::Index::Brackets { expression, .. }) => {
+                            complexity = count_expression_complexity(expression, complexity)
+                        }
+                        ast::Suffix::Call(ast::Call::AnonymousCall(
+                            ast::FunctionArgs::Parentheses { arguments, .. },
+                        )) => {
+                            for argument in arguments {
+                                complexity = count_expression_complexity(argument, complexity)
+                            }
+                        }
+                        _ => {}
+                    }
+                }
+                complexity
+            }
+
+            ast::Value::Var(ast::Var::Name(_)) => complexity,
+
+            #[cfg(feature = "roblox")]
+            ast::Value::IfExpression(if_expression) => {
+                complexity += 1;
+                if let Some(else_if_expressions) = if_expression.else_if_expressions() {
+                    for else_if_expression in else_if_expressions {
+                        complexity += 1;
+                        complexity = count_expression_complexity(
+                            else_if_expression.expression(),
+                            complexity,
+                        );
+                    }
+                }
+                complexity
+            }
+
+            _ => complexity,
+        },
+
+        _ => complexity,
+    }
+}
+
+fn count_block_complexity(block: &ast::Block, starting_complexity: u16) -> u16 {
+    let mut complexity = starting_complexity;
+
+    // we don't immediately return from the matched blocks so that we can add in any complexity from the last statement
+    for statement in block.stmts() {
+        #[cfg_attr(
+            feature = "force_exhaustive_checks",
+            deny(non_exhaustive_omitted_patterns)
+        )]
+        match statement {
+            ast::Stmt::Assignment(assignment) => {
+                for expression in assignment.expressions() {
+                    complexity = count_expression_complexity(expression, complexity);
+                }
+            }
+
+            ast::Stmt::Do(_) => {
+                // doesn't contribute dynamic branches
```

Comment:
AIUI, this might not lint something like the following
```lua
function foo()
    do
        if x then
        elseif y then
        ...
        end
    end
end
```
As nothing is traversing into the do block.
Is that meant to be handled here? If so, you may need to compute `count_block_complexity(do.block())`

Reply:
great find!